explanation/dnssec/dnssec.md

```diff
@@ -0,0 +1,254 @@
+
+ * Set the `ad` bit (Authenticated Data) in the outgoing queries.
+ * Trust the `ad` bit in the responses from the specified nameserver.
+
+When the `ad` bit is set in a DNS response, it means that DNSSEC validation was performed and successful. The data was authenticated.
+
+Specifyhing `trust-ad` in `/etc/resolv.conf` implies in these assumptions:
+
+ * The 127.0.0.53 name server is trusted to set the `ad` flag correctly in its responses. If it performs DNSSEC validation, it is trusted to perform this validation correctly, and set the `ad` flag accordingly. If it does not perform DNSSEC validation, then the `ad` flag will always be unset in the responses.
+ * The network path between localhost and 127.0.0.53 is trusted.
+
+When using systemd-resolved as the stub resolver, like in the config above, the network path is the localhost interface, and that is inherently trusted. But the actual nameserver that will be used is not 127.0.0.53: it will be whatever systemd-resolved was configured with. Unless local DNSSEC validation is enabled, systemd-resolved will remove the `ad` bit in the queries it forwards to the validating resolver, and also remove it from the responses.
+
+This is the default case in Ubuntu systems.
+
+Another valid configuration is to not use systemd-resolved, but point at the validating resolver of the network directly, like in this example:
+
+    nameserver 10.10.17.11
+    options edns0 trust-ad
+
+The `trust-ad` config means exactly the same as before: the `ad` bit will be set in outgoing queries, and its value in the responses will be trusted. But here the name server is no longer localhost, but another IP on the network. These are the assumptions being made by this config, same as before:
+
+ * The 10.10.17.11 name server is trusted to perform DNSSEC validation and set the `ad` flag accordingly in its responses.
+ * The network path between localhost and 10.10.17.11 is trusted.
+
+As these assumptions have a higher chance of not being true, this is not the default configuration.
+
+In any case, having a validating resolver in the network is a valid and very useful scenario, and good enough for most cases. And it has the extra benefit that the DNSSEC validation is done only once, at the resolver, for all clients on the network.
+
+### Local DNSSEC validation
+Some stub resolvers, like `systemd-resolved`, can perform DNSSEC validation by themselves. In this scenario, there is no insecure network path between the resolver and the dns client, as they are on the same system. But since the stub resolver will be doing the validation, that means more DNS queries, and a higher latency, because now it's not just one query, but multiple ones: a query for the actual desired record, than another query for the digital signature, then the public key, than it has to validate that key, etc. And if another client on the network happens to be making the same query, it will also have to repeat all those steps.
```

Comment:
"Some stub resolvers, such as systemd-resolved, can perform DNSSEC validation locally. This eliminates the risk of network attacks between the resolver and the client, as they reside on the same system. However, local DNSSEC validation introduces additional overhead in the form of multiple DNS queries. For each DNS query, the resolver must fetch the desired record, its digital signature, and the corresponding public key. This process can significantly increase latency, especially when multiple clients on the network request the same record."

Reply:
Adopted slightly modified.